crates/cranelift/src/func_environ.rs

```diff
@@ -746,6 +746,60 @@ impl<'module_environment> FuncEnvironment<'module_environment> {
         }
     }
 
+    fn table_index_type(&self, index: TableIndex) -> ir::Type {
+        if self.module.table_plans[index].table.table64 {
+            I64
+        } else {
+            I32
+        }
+    }
+
+    fn cast_pointer_to_table_index(
+        &self,
+        pos: &mut FuncCursor<'_>,
+        val: ir::Value,
+        index: TableIndex,
+    ) -> ir::Value {
+        let desired_type = self.table_index_type(index);
+        let pointer_type = self.pointer_type();
+        assert_eq!(pos.func.dfg.value_type(val), pointer_type);
+
+        // The current length is of type `pointer_type` but we need to fit it
+        // into `desired_type`. We are guaranteed that the result will always
+        // fit, so we just need to do the right ireduce/sextend here.
+        if pointer_type == desired_type {
+            val
+        } else if pointer_type.bits() > desired_type.bits() {
+            pos.ins().ireduce(desired_type, val)
+        } else {
+            // Note that we `sextend` instead of the probably expected
+            // `uextend`. This function is only used within the contexts of
+            // `memory.size` and `memory.grow` where we're working with units of
+            // pages instead of actual bytes, so we know that the upper bit is
+            // always cleared for "valid values". The one case we care about
+            // sextend would be when the return value of `memory.grow` is `-1`,
+            // in which case we want to copy the sign bit.
+            //
+            // This should only come up on 32-bit hosts running wasm64 modules,
+            // which at some point also makes you question various assumptions
+            // made along the way...
+            pos.ins().sextend(desired_type, val)
+        }
+    }
```

Comment:
I think this is probably copy/pasted from a memory-related function, so could the memory-related function be refactored so both tables/memories could use the same underlying logic?